File: ADT/djikstra's algo/743.js
const { PriorityQueue } = require('@datastructures-js/priority-queue');

/**
 * @param {number[][]} times
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
var networkDelayTime = function (times, n, k) {
    // Adjacency list
    const adj = new Map();
    for (const [u, v, w] of times) {
        if (!adj.has(u)) adj.set(u, []);
        adj.get(u).push([v, w]);
    }

    // Distance array
    const dist = Array(n + 1).fill(Infinity);
    dist[k] = 0;

    // Min Priority Queue (compare by distance)
    const pq = new PriorityQueue((a, b) => a[0] - b[0]);
    pq.enqueue([0, k]); // [distance, node]

    // Dijkstra
    while (!pq.isEmpty()) {
        const [currDist, node] = pq.dequeue();

        if (currDist > dist[node]) continue;

        for (const [next, weight] of (adj.get(node) || [])) {
            const newDist = currDist + weight;
            if (newDist < dist[next]) {
                dist[next] = newDist;
                pq.enqueue([newDist, next]);
            }
        }
    }

    // Result
    let ans = 0;
    for (let i = 1; i <= n; i++) {
        if (dist[i] === Infinity) return -1;
        ans = Math.max(ans, dist[i]);
    }

    return ans;
};

File: ADT/djikstra's algo/743.py
import heapq
from collections import defaultdict
from math import inf


class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adj = defaultdict(list)
        for u, v, w in times:
            adj[u].append((v, w))

        pq = []
        result = [inf] * (n + 1)
        result[k] = 0
        heapq.heappush(pq, (0, k))

        while pq:
            d, node = heapq.heappop(pq)
            for adjNode, dist in adj[node]:
                if d + dist < result[adjNode]:
                    result[adjNode] = d + dist
                    heapq.heappush(pq, (d + dist, adjNode))

        ans = max(result[1:])
        return -1 if ans == inf else ans
